src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -47,55 +49,54 @@ private class AggregatingTagger : ForegroundThreadAffinitizedObject, IAccurateTa
             private readonly object _taskGate = new object();
             private Task _taskChain;
 
-            private readonly CancellationTokenSource _initialDiagnosticsCancellationSource = new CancellationTokenSource();
+            private readonly CancellationTokenSource _shutdownCancellationSource = new CancellationTokenSource();
 
             public AggregatingTagger(
                 AbstractDiagnosticsTaggerProvider<TTag> owner,
                 ITextBuffer subjectBuffer)
             {
                 _owner = owner;
                 _subjectBuffer = subjectBuffer;
+                _taskChain = SpecializedTasks.EmptyTask;
+
+                // track context changes
+                _registration = Workspace.GetWorkspaceRegistration(_subjectBuffer.AsTextContainer());
+                _registration.WorkspaceChanged += OnWorkspaceChanged;
+
+                if (_registration.Workspace != null)
+                {
+                    // in unit test, it is possible, we create tagger for buffer which doesn't have associated document
+                    _registration.Workspace.DocumentActiveContextChanged += OnDocumentActiveContextChanged;
+                }
 
                 var document = _subjectBuffer.AsTextContainer().GetOpenDocumentInCurrentContext();
                 _currentDocumentId = document?.Id;
 
-                // Kick off a background task to collect the initial set of diagnostics.
-                var cancellationToken = _initialDiagnosticsCancellationSource.Token;
-                var asyncToken = _owner._listener.BeginAsyncOperation(GetType() + ".GetInitialDiagnostics");
-                var task = Task.Run(() => GetInitialDiagnosticsInBackground(document, cancellationToken), cancellationToken);
-                task.CompletesAsyncOperation(asyncToken);
-
-                _taskChain = task;
+                // make sure we belong to same workspace
+                Contract.ThrowIfFalse(document == null || _registration.Workspace == document.Project.Solution.Workspace);
```

Comment:
My first reaction was "that could never happen" and this could be deleted. But then again I guess that's why to leave it... :-)